File: Sources/CodeRim/Sources/ClaudeJSONLParser.swift
import Foundation

struct ClaudeUsageObservation: Sendable {
    let messageID: String
    let sessionID: String
    let agentID: String?
    let occurredAt: Date
    let model: String
    let workingDirectory: String?
    let usage: TokenUsage
}

/// Projects only accounting fields. Message content and credentials are never decoded.
struct ClaudeJSONLParser: Sendable {
    private struct Row: Decodable {
        let type: String
        let timestamp: String
        let sessionId: String
        let agentId: String?
        let cwd: String?
        let isApiErrorMessage: Bool?
        let message: Message
    }

    private struct Message: Decodable {
        let id: String
        let model: String
        let role: String
        let usage: Usage
    }

    private struct Usage: Decodable {
        let input_tokens: Int64
        let cache_read_input_tokens: Int64?
        let cache_creation_input_tokens: Int64?
        let output_tokens: Int64
    }

    func parse(_ line: Data) -> ClaudeUsageObservation? {
        guard line.count <= CodexJSONLParser.maximumLineBytes,
              let row = try? JSONDecoder().decode(Row.self, from: line),
              row.type == "assistant", row.message.role == "assistant",
              row.isApiErrorMessage != true,
              Self.validIdentifier(row.message.id), Self.validIdentifier(row.sessionId),
              Self.validIdentifier(row.message.model), row.message.model.hasPrefix("claude-"),
              let timestamp = Self.date(row.timestamp)
        else { return nil }

        let raw = row.message.usage
        let read = raw.cache_read_input_tokens ?? 0
        let write = raw.cache_creation_input_tokens ?? 0
        guard [raw.input_tokens, read, write, raw.output_tokens].allSatisfy({
            (0...1_000_000_000_000).contains($0)
        }) else { return nil }

        // Unlike Codex, Claude's input_tokens excludes both cache reads and writes.
        // Nested TTL, iteration and thinking fields are breakdowns, not extra tokens.
        let usage = TokenUsage(
            inputTokens: raw.input_tokens + read + write,
            cachedInputTokens: read,
            cacheWriteInputTokens: write,
            outputTokens: raw.output_tokens
        )
        return ClaudeUsageObservation(
            messageID: row.message.id,
            sessionID: row.sessionId,
            agentID: row.agentId.flatMap { Self.validIdentifier($0) ? $0 : nil },
            occurredAt: timestamp,
            model: row.message.model.lowercased(),
            workingDirectory: row.cwd,
            usage: usage
        )
    }

    private static func validIdentifier(_ value: String) -> Bool {
        !value.isEmpty && value.utf8.count <= 128 && value.utf8.allSatisfy {
            (48...57).contains($0) || (65...90).contains($0)
                || (97...122).contains($0) || [45, 46, 95].contains($0)
        }
    }

    private static func date(_ value: String) -> Date? {
        guard value.utf8.count <= 64 else { return nil }
        return (try? Date(value, strategy: .iso8601.time(includingFractionalSeconds: true)))
            ?? (try? Date(value, strategy: .iso8601))
    }
}

File: Windows/src/CodeRim.Core/Parsing/ClaudeJsonlParser.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using CodeRim.Core.Domain;

namespace CodeRim.Core.Parsing;

/// Projects only accounting metadata. Prompts and response bodies never leave the parser.
public static class ClaudeJsonlParser
{
    public static UsageEvent? Parse(ReadOnlyMemory<byte> line, byte[]? projectKey = null, string? sourceName = null)
    {
        if (line.Length > CodexJsonlParser.MaximumLineBytes) return null;
        try
        {
            using var document = JsonDocument.Parse(line);
            var root = document.RootElement;
            if (JsonFields.Text(root, "type") != "assistant" || JsonFields.Flag(root, "isApiErrorMessage")) return null;
            var message = JsonFields.Object(root, "message");
            if (JsonFields.Text(message, "role") != "assistant") return null;
            var id = JsonFields.Text(message, "id");
            var session = JsonFields.Text(root, "sessionId");
            var model = JsonFields.Text(message, "model");
            if (id is null || session is null || !ValidIdentifier(session) || model is null || !model.StartsWith("claude-", StringComparison.Ordinal)) return null;
            if (!DateTimeOffset.TryParse(JsonFields.Text(root, "timestamp"), System.Globalization.CultureInfo.InvariantCulture,
                    System.Globalization.DateTimeStyles.AssumeUniversal, out var date)) return null;
            var usage = JsonFields.Object(message, "usage");
            var input = JsonFields.Count(usage, "input_tokens");
            var output = JsonFields.Count(usage, "output_tokens");
            var read = JsonFields.OptionalCount(usage, "cache_read_input_tokens");
            var write = JsonFields.OptionalCount(usage, "cache_creation_input_tokens");
            if (input is null || output is null || read is null || write is null) return null;
            var agent = AgentIdentifier(JsonFields.Text(root, "agentId"));
            if (agent is null && sourceName?.StartsWith("agent-", StringComparison.Ordinal) == true) agent = AgentIdentifier(sourceName);
            var parent = Hash(session);
            var owner = agent is null ? parent : Hash("claude-agent|" + session + "|" + agent);
            return new UsageEvent(Hash(id), date, new TokenUsage(input.Value + read.Value + write.Value, read.Value, output.Value, write.Value),
                model.ToLowerInvariant(), JsonFields.Folder(JsonFields.Text(root, "cwd")), owner, "claude", ProjectIdentity(JsonFields.Text(root, "cwd"), projectKey))
                { ImportParentSessionId = agent is null ? null : parent };
        }
        catch (JsonException) { return null; }
    }

    // Stream revisions take maxima of disjoint components, never max(total input).
    public static UsageEvent Merge(UsageEvent first, UsageEvent next)
    {
        var read = Math.Max(first.Usage.CachedInputTokens, next.Usage.CachedInputTokens);
        var write = Math.Max(first.Usage.CacheWriteInputTokens ?? 0, next.Usage.CacheWriteInputTokens ?? 0);
        var input = Math.Max(first.Usage.UncachedInputTokens, next.Usage.UncachedInputTokens);
        // An agent transcript can also occur in a copied parent history. Explicit
        // agent ownership wins only over that exact parent, never an unrelated task.
        var owner = first.ImportParentSessionId is null && next.ImportParentSessionId == first.SessionId ? next : first;
        return owner with
        {
            OccurredAt = first.OccurredAt < next.OccurredAt ? first.OccurredAt : next.OccurredAt,
            Usage = new TokenUsage(input + read + write, read, Math.Max(first.Usage.OutputTokens, next.Usage.OutputTokens), write)
        };
    }

    private static string? AgentIdentifier(string? value)
    {
        if (value is null || !ValidIdentifier(value)) return null;
        return value.StartsWith("agent-", StringComparison.Ordinal) ? value : "agent-" + value;
    }

    private static bool ValidIdentifier(string value) => value is { Length: > 0 and <= 128 }
        && value.All(c => char.IsAsciiLetterOrDigit(c) || c is '-' or '.' or '_');

    private static readonly byte[] ProcessProjectKey = RandomNumberGenerator.GetBytes(32);
    internal static string ProjectIdentity(string? path, byte[]? key) => path is null ? "unknown" :
        Convert.ToHexString(HMACSHA256.HashData(key ?? ProcessProjectKey, Encoding.UTF8.GetBytes(path.Replace('\\', '/').TrimEnd('/'))));
    internal static string Hash(string value) => Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(value)));
}

internal static class JsonFields
{
    internal static JsonElement Object(JsonElement parent, string name) =>
        parent.ValueKind == JsonValueKind.Object && parent.TryGetProperty(name, out var value) ? value : default;
    internal static string? Text(JsonElement parent, string name)
    {
        var value = Object(parent, name);
        if (value.ValueKind != JsonValueKind.String) return null;
        var text = value.GetString();
        return text is { Length: > 0 and <= 512 } && !text.Any(char.IsControl) ? text : null;
    }
    internal static bool Flag(JsonElement parent, string name) => Object(parent, name).ValueKind == JsonValueKind.True;
    internal static long? Count(JsonElement parent, string name)
    {
        var value = Object(parent, name);
        return value.ValueKind == JsonValueKind.Number && value.TryGetInt64(out var number)
            && number is >= 0 and <= CodexJsonlParser.MaximumTokenComponent ? number : null;
    }
    internal static long? OptionalCount(JsonElement parent, string name) =>
        Object(parent, name).ValueKind == JsonValueKind.Undefined ? 0 : Count(parent, name);
    internal static string Folder(string? path) => path is null ? "Unknown project" : path.Replace('\\', '/').TrimEnd('/').Split('/').Last();
}
